README
= Description
  A Ruby interface for getting file system information.

= Installation
  gem install sys-filesystem

= Synopsis
  require 'sys/filesystem'
  include Sys
   
  # Display information about a particular filesystem.
  p Filesystem.stat('/')

  # Sample output

  #<Sys::Filesystem::Stat:0x517440
    @base_type = "ufs",
    @flags = 4,
    @files_available = 3817457,
    @block_size = 8192,
    @blocks_available = 19957633,
    @blocks = 34349612,
    @name_max = 255,
    @path = "/",
    @filesystem_id = 35651592,
    @files = 4135040,
    @fragment_size = 1024,
    @files_free = 3817457,
    @blocks_free = 20301129
  >
   
  # Describe all mount points on the system
  Filesystem.mounts{ |mount| p mount }
   
  # Find the mount point of any particular file
  puts Filesystem.mount_point('/home/djberge/some_file.txt') => '/home'

= Notes
  This is a pure Ruby implementation that uses FFI. This means it should work
  with JRuby, too.

= Sample code
  Run 'rake example' if you want to see a basic sample run. The actual code
  is 'example_stat.rb' in the 'examples' directory. Modify it as you see fit.

= Known Bugs
  None that I'm aware of. Please report bugs on the project page at
  https://github.com/djberg96/sys-filesystem

= Future Plans
  Add better 64-bit support for Linux and BSD.
  Other suggestions welcome.

= Acknowledgements
  Mike Hall, for ideas and code that I borrowed from his 'filesystem' library.
   
  Park Heesob, for implementation and API ideas for the MS Windows version.
   
  Nobuyoshi Miyokawa, for adding the original FreeBSD and OS X support.
   
= License
  Apache-2.0

= Copyright
  (C) 2003-2019 Daniel J. Berger
  All Rights Reserved

= Warranty
  This library is provided "as is" and without any express or
  implied warranties, including, without limitation, the implied
  warranties of merchantability and fitness for a particular purpose.

= Author
  Daniel J. Berger


CHANGES
== 1.3.0 - 3-Oct-2019
* Added the mount and umount singleton methods.
* Changed an internal class variable to a frozen constant.

== 1.2.0 - 1-Jan-2019
* Changed the license to Apache-2.0.
* Added the case_sensitive? and case_insensitive? instance methods to the
  Filesystem::Stat class.
* Refactored the Numeric helper methods into a single location.
* Fixed the deprecation warnings when running the tests on Windows.
* Added metadata to the gemspec.

== 1.1.9 - 26-Jan-2018
* The bytes_free and bytes_total methods were updated to use the fragment size
  instead of the block size. This fixes Macs using APFS, as well as Solaris 10.
  Thanks go to Mikhail Vasin for the spot.
* Removed the code specific to FreeBSD 10, which was ultimately an attempt
  to address the issue now resolved by the above change.
* Updated the cert again.

== 1.1.8 - 2-Sep-2017
* The stat and mount_point methods now call to_s on the string argument
  explicitly so that Pathname objects will now work with String#wincode.
* Updated some tests to avoid Fixnum warnings for Ruby 2.4.
* Updated the cert.

== 1.1.7 - 1-Aug-2016
* Fixed an issue in the stat method for MS Windows where the blocks_avail
  value was not correct. Thanks go to Maxime Lapointe for the spot.

== 1.1.6 - 17-May-2016
* On versions that use setmntent or fopen, explicitly raise an error if
  either of those calls return a null value.

== 1.1.5 - 5-Dec-2015
* This gem is now signed.
* The gem related tasks in the Rakefile now assume Rubygems 2.x.
* Updates to the gemspec, added cert_chain, removed rubyforge_project.
* Internal reorganization, and use of relative_require as appropriate.
* Added a sys-filesystem.rb file for convenience.

== 1.1.4 - 15-Mar-2015
* The File.mounts method no longer raises an error if a mount point
  is not accessible. Thanks go to Michael Pope for the patch.
* Some internal code reorganization.

== 1.1.3 - 1-Oct-2014
* Now ignores EPERM errors when trying to find the mount point for a path.
  Thanks go to petersen for the patch.
* The Filesystem.stat method now defaults to using the root path on Windows
  for volume information.

== 1.1.2 - 9-May-2014
* Added the percent_used, bytes_total, bytes_free and bytes_used methods
  to the Filesystem::Stat class. Thanks go to xanview for the suggestion.
* Changed File.exists? to File.exist? to avoid warnings in Ruby 2.1.x.
* The convenience methods to_mb, to_gb, etc, are now defined in Numeric
  instead of Fixnum.
* Added the to_tb method for terabytes.
* Minor updates to the gem:create and gem:install Rake tasks.

== 1.1.1 - 3-Oct-2013
* Solaris now always uses statvfs64 on Solaris for better 64-bit support.
  Thanks go to Jeff Shantz for the spot.

== 1.1.0 - 19-Jan-2013
* Converted the Windows source code to use FFI. Consequently, there is now
  a single gem rather than separate gems for Windows and Unix.
* Revamped the Windows tests.

== 1.0.0 - 11-Jan-2012
* Converted everything from C to FFI for the Unix flavors. The Windows
  source code remains untouched.

== 0.3.4 - 19-Nov-2010
* Fixed a bug where negative block counts were happening on very large
  hard drives. Thanks go to Jonas Pfenniger for the spot.
* Refactored the clean task in the Rakefile.
* Some cosmetic source code changes.

== 0.3.3 - 21-May-2010
* Added a workaround for the Sys::Filesystem#block_size member to deal with
  a bug in OS X. Thanks go to Josh Pasqualetto for the spot.

== 0.3.2 - 29-Dec-2009
* Source has been moved to github.
* Added the 'gem' task and removed build logic from the gemspec.
* Updated the install task.
* Minor correction to the manifest.
* Removed some junk build files that were inadvertently included in
  the last gem.

== 0.3.1 - 5-Aug-2009
* Now compatible with Ruby 1.9.x.
* Changed license to Artistic 2.0
* Updated the gemspec, including the explicit addition of a license and
  test-unit as a development dependency, as well as an updated description. 

== 0.3.0 - 26-Feb-2009
* Added support for OS X and FreeBSD thanks to an awesome patch by Nobuyoshi
  Miyokawa.
* Added the Filesystem.mount_point method that takes a file and returns
  the mount point it's sitting on.

== 0.2.0 - 30-Dec-2008
* Added the Filesystem.mounts method for iterating over mount or volume
  information.

== 0.1.1 - 28-Mar-2007
* Bug fix for BSD flavors.  Thanks go to Jeremy Kemper and Ole Christian
  Rynning for the spot.
* Bug fix for OS X (along the same lines as the BSD fix). Thanks go to
  Aurelian Dehay for the spot.
* Some Rdoc improvements for the C extension.
* Tweaks to the gemspec.
* Added synopsis to the README.

== 0.1.0 - 17-Nov-2006
* Initial release.  Alpha. Code is stable, but API is not.


ext/sys/filesystem.c
/********************************************************************
 * filesystem.c
 *
 * This is the source file for the sys-filesystem package for Ruby.
 * This is for UNIX platforms only.
 ********************************************************************/
#include <ruby.h>

#include <sys/types.h>

#ifdef HAVE_SYS_STATVFS_H // Most flavors of UNIX
#include <sys/statvfs.h>
#else                     // FreeBSD 4.x
#include <sys/param.h>
#include <sys/mount.h>
#include <sys/vnode.h>
#endif

#ifdef HAVE_SYS_MNTTAB_H  // Solaris

#include <sys/mnttab.h>
#define MNTENT mnttab
#define START_MNT(F,M) fopen(F,M)
#define GET_MNT(FP,MP) (getmntent(FP,MP) == 0)
#define END_MNT(F) fclose(F)
#define MOUNTFILE "/etc/mnttab"

// Ruby 1.9.x compatibility
#ifndef RSTRING_PTR
#define RSTRING_PTR(v) (RSTRING(v)->ptr)
#define RSTRING_LEN(v) (RSTRING(v)->len)
#endif

#ifndef RARRAY_PTR
#define RARRAY_PTR(v) (RARRAY(v)->ptr)
#define RARRAY_LEN(v) (RARRAY(v)->len)
#endif

#elif HAVE_GETMNTINFO

struct _ment {
  struct statfs *mntbufp;
  int current;
  int max;
};

#include <sys/param.h>
#include <sys/mount.h>
#include <sys/vnode.h>

#define MNTENT _ment
#define START_MNT(F,M) start_mnt(F,M)
#define GET_MNT(FP,MP) ((MP = get_mnt(MP)) != NULL)
#define END_MNT(F) end_mnt(F)
#define MOUNTFILE "getmntinfo"

#else // Most flavors of UNIX

#ifdef HAVE_MNTENT_H
#include <mntent.h>
#define MNTENT mntent
#define START_MNT(F,M) setmntent(F,M)
#define GET_MNT(FP,MP) ((MP = getmntent(FP)) != NULL)
#define END_MNT(F) endmntent(F)
#define MOUNTFILE "/etc/mtab"
#endif

#endif

#ifdef HAVE_GETMNTINFO

// This table comes from FreeBSD mount.c@1.105
static struct opt {
  int o_opt;
  const char *o_name;
} optnames[] = {
  { MNT_ASYNC,    "asynchronous" },
  { MNT_EXPORTED, "NFS exported" },
  { MNT_LOCAL,    "local" },
  { MNT_NOEXEC,      "noexec" },
  { MNT_NOSUID,      "nosuid" },
  { MNT_QUOTA,    "with quotas" },
  { MNT_RDONLY,      "read-only" },
  { MNT_SYNCHRONOUS, "synchronous" },
  { MNT_UNION,    "union" },
#ifdef HAVE_MNT_MULTILABEL
  { MNT_MULTILABEL,  "multilabel" },
#endif
#ifdef HAVE_MNT_NOATIME
  { MNT_NOATIME,  "noatime" },
#endif
#if !defined(__MACH__) || !defined(__APPLE__)
  { MNT_NOSYMFOLLOW, "nosymfollow" },
  { MNT_NOCLUSTERR,  "noclusterr" },
  { MNT_NOCLUSTERW,  "noclusterw" },
  { MNT_SUIDDIR,  "suiddir" },
  { MNT_SOFTDEP,  "soft-updates" },
  { MNT_ACLS,     "acls" },
#endif
  { 0, NULL }
};

static FILE* start_mnt(const char *filename, const char *type)
{
  return (FILE*)!0; // Do nothing
}

static struct _ment* get_mnt(struct _ment* m)
{
  struct _ment* ret = m;

  if (m->max == 0) {
    if ((m->max = getmntinfo(&(m->mntbufp), MNT_NOWAIT)) == 0) {
      return NULL;
    }
    m->current = 0;
  }

  if (m->current >= m->max) {
    ret = NULL;
  }
  m->current++;

  return ret;
}

static void end_mnt(FILE* fp)
{
  // Do nothing
}

#endif


VALUE mSys, cFilesys, cStat, cMount;

static VALUE create_mount_object(struct MNTENT*);

/* call-seq:
 *    Filesystem.stat(path)
 *
 * Returns a a Filesystem::Stat object containing information about the +path+
 * file system.
 */
static VALUE fs_stat(VALUE klass, VALUE v_path){
  VALUE v_stat;
  char* path = StringValuePtr(v_path);

#ifdef HAVE_STATVFS
  struct statvfs fs;

  if(statvfs(path, &fs) < 0)
    rb_sys_fail("statvfs");
#else
  struct mount mp;
  struct statfs fs;
  struct proc p;

  if(VFS_STATFS(&mp, &fs, &p) < 0)
    rb_sys_fail("VFS_STATFS");
#endif

  v_stat = rb_funcall(cStat, rb_intern("new"), 0, 0);

  rb_iv_set(v_stat, "@path", v_path);

// You gotta love OS X, right?
#ifdef __MACH__
  rb_iv_set(v_stat, "@block_size", ULONG2NUM(fs.f_bsize/256));
#else
  rb_iv_set(v_stat, "@block_size", ULONG2NUM(fs.f_bsize));
#endif

  rb_iv_set(v_stat, "@fragment_size", ULONG2NUM(fs.f_frsize));
  rb_iv_set(v_stat, "@blocks", ULONG2NUM(fs.f_blocks));
  rb_iv_set(v_stat, "@blocks_free", ULONG2NUM(fs.f_bfree));
  rb_iv_set(v_stat, "@blocks_available", ULONG2NUM(fs.f_bavail));
  rb_iv_set(v_stat, "@files", ULONG2NUM(fs.f_files));
  rb_iv_set(v_stat, "@files_free", ULONG2NUM(fs.f_ffree));
  rb_iv_set(v_stat, "@files_available", ULONG2NUM(fs.f_favail));
  rb_iv_set(v_stat, "@filesystem_id", ULONG2NUM(fs.f_fsid));
  rb_iv_set(v_stat, "@flags", ULONG2NUM(fs.f_flag));
  rb_iv_set(v_stat, "@name_max", ULONG2NUM(fs.f_namemax));

#ifdef HAVE_ST_F_BASETYPE
  rb_iv_set(v_stat, "@base_type", rb_str_new2(fs.f_basetype));
#endif

  return rb_obj_freeze(v_stat);
}

/* Convenient methods for converting bytes to kilobytes, megabytes or
 * gigabytes.
 */

/*
 * call-seq:
 *    <tt>fix</tt>.to_kb
 *
 * Returns +fix+ in terms of kilobytes.
 */
static VALUE fixnum_to_kb(VALUE self){
  return ULL2NUM(NUM2ULONG(self) / 1024);
}

/*
 * call-seq:
 *    <tt>fix</tt>.to_mb
 *
 * Returns +fix+ in terms of megabytes.
 */
static VALUE fixnum_to_mb(VALUE self){
  return ULL2NUM(NUM2ULONG(self) / 1048576);
}

/*
 * call-seq:
 *    <tt>fix</tt>.to_gb
 *
 * Returns +fix+ in terms of gigabytes.
 */
static VALUE fixnum_to_gb(VALUE self){
  return ULL2NUM(NUM2ULONG(self) / 1073741824);
}

/*
 * call-seq:
 *    Filesystem.mounts
 *    Filesystem.mounts{ ... }
 *
 * In block form, yields a Filesystem::Mount object for each mounted filesystem
 * on your machine. In non-block form, returns an array of Filesystem::Mount
 * objects instead.
 *
 * Example:
 *
 *    Filesystem.mounts{ |fs|
 *       p fs.name            # => e.g. '/dev/dsk/c0t0d0s0', 'proc', etc
 *       p fs.mount_time      # => e.g. Thu Dec 11 15:07:23 -0700 2008
 *       p fs.mount_type      # => e.g. 'ufs', 'proc', etc
 *       p fs.mount_point     # => e.g. '/', '/proc', '/tmp', etc
 *       p fs.options         # => e.g. "rw,intr,largefiles,logging,xattr,onerror=panic,dev=2200008"
 *       p fs.pass_number     # => e.g. ???
 *       p fs.dump_frequency  # => e.g. ???
 *    }
 */
static VALUE fs_mounts(VALUE klass){
   VALUE v_array;
   FILE* fp;
   struct MNTENT* mp;
#ifdef HAVE_SYS_MNTTAB_H
   struct MNTENT mt;
   mp = &mt;
#elif HAVE_GETMNTINFO
   struct MNTENT mt;
   mt.max = 0;
   mp = &mt;
#endif

   v_array = Qnil;

   if((fp = START_MNT(MOUNTFILE, "r")) == NULL)
      rb_sys_fail(MOUNTFILE);

   if(rb_block_given_p()){
      while(GET_MNT(fp, mp))
         rb_yield(create_mount_object(mp));
   }
   else{
      v_array = rb_ary_new();
      while(GET_MNT(fp, mp))
         rb_ary_push(v_array, create_mount_object(mp));
   }

   END_MNT(fp);

   return v_array; /* nil in block form */
}

/* Private function to create a Filesystem object */
static VALUE create_mount_object(struct MNTENT* mp){
   VALUE v_mount = rb_funcall(cMount, rb_intern("new"), 0, 0);

#ifdef HAVE_SYS_MNTTAB_H
   rb_iv_set(v_mount, "@name", rb_tainted_str_new2(mp->mnt_special));
   rb_iv_set(v_mount, "@mount_point", rb_tainted_str_new2(mp->mnt_mountp));
   rb_iv_set(v_mount, "@mount_type", rb_tainted_str_new2(mp->mnt_fstype));
   rb_iv_set(v_mount, "@options", rb_tainted_str_new2(mp->mnt_mntopts));
   rb_iv_set(v_mount, "@mount_time", rb_time_new(atoi(mp->mnt_time), 0));
   rb_iv_set(v_mount, "@dump_frequency", Qnil);
   rb_iv_set(v_mount, "@pass_number", Qnil);
#elif HAVE_GETMNTINFO
   {
      struct statfs *p = mp->mntbufp + (mp->current-1);
      struct opt *o;
      int flags, mul;
      char ostr[BUFSIZ];

      flags = p->f_flags & MNT_VISFLAGMASK;
      ostr[0] = '\0';

      for (mul = 0, o = optnames; flags && o->o_opt; o++) {
         if (flags & o->o_opt) {
            strlcat(ostr, ((mul++) ? "," : ""), BUFSIZ);
            strlcat(ostr, o->o_name, BUFSIZ);
            flags &= ~o->o_opt;
         }
      }

      rb_iv_set(v_mount, "@name", rb_tainted_str_new2(p->f_mntfromname));
      rb_iv_set(v_mount, "@mount_point", rb_tainted_str_new2(p->f_mntonname));
      rb_iv_set(v_mount, "@mount_type", rb_tainted_str_new2(p->f_fstypename));
      rb_iv_set(v_mount, "@options", rb_tainted_str_new2(ostr));
      rb_iv_set(v_mount, "@mount_time", Qnil);
      rb_iv_set(v_mount, "@dump_frequency", Qnil);
      rb_iv_set(v_mount, "@pass_number", Qnil);
   }
#else
   rb_iv_set(v_mount, "@name", rb_tainted_str_new2(mp->mnt_fsname));
   rb_iv_set(v_mount, "@mount_point", rb_tainted_str_new2(mp->mnt_dir));
   rb_iv_set(v_mount, "@mount_type", rb_tainted_str_new2(mp->mnt_type));
   rb_iv_set(v_mount, "@options", rb_tainted_str_new2(mp->mnt_opts));
   rb_iv_set(v_mount, "@mount_time", Qnil);
   rb_iv_set(v_mount, "@dump_frequency", INT2NUM(mp->mnt_freq));
   rb_iv_set(v_mount, "@pass_number", INT2NUM(mp->mnt_passno));
#endif

   return v_mount;
}

/*
 * call-seq:
 *    Filesystem.mount_point(file)
 *
 * Returns the mount point of the given +file+, or itself if it cannot be
 * determined.
 *
 * Example:
 *
 *    Filesystem.mount_point('/home/djberge/some_file.txt') => '/home'
 */
static VALUE fs_mount_point(VALUE klass, VALUE v_file){
   VALUE v_stat, v_stat_m, v_mounts, v_mount_pt, v_mount;
   VALUE v_found = Qfalse;
   long dev1, dev2;
   int i = 0;

   v_stat   = rb_funcall(rb_cFile, rb_intern("stat"), 1, v_file);
   v_mounts = fs_mounts(klass);
   dev1     = FIX2LONG(rb_funcall(v_stat, rb_intern("dev"), 0, 0));

   /* Stat each mount point and compare its device number with the device
    * number of the file provided. If they match, we have a winner.
    */
   for(i = 0; i < RARRAY_LEN(v_mounts); i++){
      v_mount    = RARRAY_PTR(v_mounts)[i];
      v_mount_pt = rb_funcall(v_mount, rb_intern("mount_point"), 0, 0);
      v_stat_m   = rb_funcall(rb_cFile, rb_intern("stat"), 1, v_mount_pt);
      dev2       = FIX2LONG(rb_funcall(v_stat_m, rb_intern("dev"), 0, 0));

      if(dev1 == dev2){
         v_found = Qtrue;
         break;
      }
   }

   if(v_found == Qtrue)
      return v_mount_pt;
   else
      return v_file;
}

void Init_filesystem(){
  /* The toplevel namespace */
  mSys = rb_define_module("Sys");

  /* The Filesystem class serves an abstract base class. It's methods return
   * objects of other types. Do not instantiate.
   */
  cFilesys = rb_define_class_under(mSys, "Filesystem", rb_cObject);

  /* Instances of this class are returned by the Filesystem.mount method */
  cMount = rb_define_class_under(cFilesys, "Mount", rb_cObject);

  /* Instances of this class are returned by the Filesystem.stat method */
  cStat = rb_define_class_under(cFilesys, "Stat", rb_cObject);

  /* Singleton methods */
  rb_define_singleton_method(cFilesys, "mount_point", fs_mount_point, 1);
  rb_define_singleton_method(cFilesys, "mounts", fs_mounts, 0);
  rb_define_singleton_method(cFilesys, "stat", fs_stat, 1);

  /* Filesystem::Mount accessors */

  /* The name of the mounted resource */
  rb_define_attr(cMount, "name", 1, 0);

  /* The mount point/directory */
  rb_define_attr(cMount, "mount_point", 1, 0);

  /* The type of the file system mount, e.g. 'ufs', 'nfs', etc */
  rb_define_attr(cMount, "mount_type", 1, 0);

  /* A list of comma separated options for the mount, e.g. 'rw', etc */
  rb_define_attr(cMount, "options", 1, 0);

  /* The time the file system was mounted or nil if not supported */
  rb_define_attr(cMount, "mount_time", 1, 0);

  /* The dump frequency in days (or nil if not supported) */
  rb_define_attr(cMount, "dump_frequency", 1, 0);

  /* The pass number of the file system check or nil if not supported */
  rb_define_attr(cMount, "pass_number", 1, 0);

  /* Filesystem::Mount Aliases */

  rb_define_alias(cMount, "fsname", "name");
  rb_define_alias(cMount, "dir", "mount_point");
  rb_define_alias(cMount, "opts", "options");
  rb_define_alias(cMount, "passno", "pass_number");
  rb_define_alias(cMount, "freq", "dump_frequency");

  /* Filesystem::Stat accessors */

  /* The path of the file system */
  rb_define_attr(cStat, "path", 1, 0);

  /* The preferred system block size */
  rb_define_attr(cStat, "block_size", 1, 0);

  /* The fragment size, i.e. fundamental file system block size */
  rb_define_attr(cStat, "fragment_size", 1, 0);

  /* The total number of +fragment_size+ blocks in the file system */
  rb_define_attr(cStat, "blocks", 1, 0);

  /* The total number of free blocks in the file system */
  rb_define_attr(cStat, "blocks_free", 1, 0);

  /* The number of free blocks available to unprivileged processes */
  rb_define_attr(cStat, "blocks_available", 1, 0);

  /* The total number of files/inodes that can be created */
  rb_define_attr(cStat, "files", 1, 0);

  /* The total number of free files/inodes on the file system */
  rb_define_attr(cStat, "files_free", 1, 0);

  /* The number of free files/inodes available to unprivileged processes */
  rb_define_attr(cStat, "files_available", 1, 0);

  /* The file system identifier */
  rb_define_attr(cStat, "filesystem_id", 1, 0);

  /* The file system type, e.g. UFS */
  rb_define_attr(cStat, "base_type", 1, 0);

  /* A bit mask of flags.  See the <tt>Constants</tt> for a list of flags */
  rb_define_attr(cStat, "flags", 1, 0);

  /* The maximum length of a file name permitted on the file system */
  rb_define_attr(cStat, "name_max", 1, 0);

  // Constants

  /* 0.3.4: The version of this library (a String) */
  rb_define_const(cFilesys, "VERSION", rb_str_new2("0.3.4"));

  /* 0x00000001: Read only file system */
  rb_define_const(cStat, "RDONLY", INT2FIX(ST_RDONLY));

  /* 0x00000002: File system does not support suid or sgid semantics */
  rb_define_const(cStat, "NOSUID", INT2FIX(ST_NOSUID));

#ifdef ST_NOTRUNC
  /* 0x00000003: File system does not truncate file names longer than +name_max+ */
  rb_define_const(cStat, "NOTRUNC", INT2FIX(ST_NOTRUNC));
#endif

  // Aliases

  rb_define_alias(cStat, "inodes", "files");
  rb_define_alias(cStat, "inodes_free", "files_free");
  rb_define_alias(cStat, "inodes_available", "files_available");

  // Convenient methods for Fixnum

  rb_define_method(rb_cFixnum, "to_kb", fixnum_to_kb, 0);
  rb_define_method(rb_cFixnum, "to_mb", fixnum_to_mb, 0);
  rb_define_method(rb_cFixnum, "to_gb", fixnum_to_gb, 0);
}
